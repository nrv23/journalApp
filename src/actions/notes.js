import { db } from '../firebase/firebaseConfig';
import { loadNotes } from '../helpers/loadNotes';
import { types } from '../types';

export const starNewNote = () =>  {

    return async (dispatch,getState) => { // el segundo parÃ¡metro del callback que devuelve la funcion asincrona, es una funcion
        // que devuelve todo el state global actual
        try {
            
            const { auth:{uid}  } = getState();

            const newNote = {

                title: '',
                body: '',
                date: new Date().getTime()
            }

            const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);
            dispatch(activeNote(docRef.id,newNote));
        } catch (error) {
            console.log(error)
        }
    }
}

export const activeNote = (id,note) => {

    return {
        type: types.NOTESACTIVE,
        payload: {
            ...note,
            id
        }
    }
}

export const startLoadingNotes = (uid) => {

    return async dispatch => { // redux-thunk permite la ejecucion de dispatchs asincronos
        const notes = await loadNotes(uid);

        dispatch(setNotes(notes));
    }
}

export const setNotes = notes => ({
    type: types.NOTESLOAD,
    payload: notes
})