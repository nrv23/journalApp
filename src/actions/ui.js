import { types } from "../types"



export const setError = (error) => {
    
    return {
        type: types.UISETERROR,
        payload: error
    }
}

export const unSetError = () => {
    
    return {
        type: types.UIUNSETERROR
    }
}

export const startLoading = () => {

    return {
        type: types.UISTARTLOADING,
        payload: true
    }
}


export const finishLoading = () => {

    return {
        type: types.UISTARTLOADING,
        payload: false
    }
}