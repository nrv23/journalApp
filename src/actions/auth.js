import  { firebase,googleAuthProvider } from '../firebase/firebaseConfig';
import { types } from '../types/index';
import { clearNotes } from './notes';
import { startLoading,finishLoading, setError,unSetError } from './ui';


export const startLoginEmailPassword = (email, pass) => {
    
    return dispatch => { // retornar un callback para ejecutar un action asincrono
        dispatch(startLoading());
        firebase.auth().signInWithEmailAndPassword(email,pass).then(({user:{uid,displayName}}) => {
            dispatch(login(uid,displayName))
            dispatch(finishLoading());
            dispatch(unSetError());
        })
        .catch(e => {
            dispatch(finishLoading());
            console.log(e.code)

            if(e.code === 'auth/invalid-email'|| e.code === 'auth/wrong-password') { 
                dispatch(setError('Credenciales incorrectos'))
            } else {

                dispatch(setError("Hubo un error en el login"));
            }

        })
    }
}

export const login = (uid, displayName) => {

    return {
        type: types.LOGIN,
        payload: {
            uid,
            displayName
        }
    }
}

export const startGoogleLogin = () => {


    return dispatch => {
        firebase.auth().signInWithPopup(googleAuthProvider)
        .then(({user:{uid,displayName}}) => {
            dispatch(login(uid,displayName))
        })
    }
}

export const startRegisterWithEmailPasswordName = (email,password,name) => {

    return dispatch => {

        firebase.auth().createUserWithEmailAndPassword(email,password)
            .then(async ({user}) => {

                await user.updateProfile({displayName: name});
                dispatch(login(user.uid,user.displayName))
            })
            .catch(e => {
                console.log(e.code)
            })
    }

}

export const startLogout = () => {

    return async dispatch => {

        await firebase.auth().signOut();
        dispatch(logout());
        dispatch(clearNotes());
    }
}

export const logout = () => {

    return {
        type: types.LOGOUT
    }
}