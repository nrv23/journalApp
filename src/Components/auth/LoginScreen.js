import React from 'react'
import { Link } from 'react-router-dom';
import useForm from '../../hooks/useForm';
import  { useDispatch,useSelector  } from 'react-redux';
import { startLoginEmailPassword,startGoogleLogin } from '../../actions/auth';


export const LoginScreen = () => {


    const dispatch = useDispatch();

    const {loading,msgError} = useSelector(state=>state.ui);
    console.log({msgError})
    const [formValues,handleInputChange] = useForm({
        email: '',
        password: ''
    });

    const handleSubmit = e => {
        
        e.preventDefault();
       
        
        dispatch(startLoginEmailPassword(email,password))
  
    }

    const handleGoogleSigning = () => {
        dispatch(startGoogleLogin())
    }

    const { email, password } = formValues;
    return (
        <>
            <h3 className="auth__title">Login</h3>

            <form
                onSubmit={handleSubmit}
            >

                {
                    msgError && <div className="alert alert-danger">{msgError}</div>
                }

                <input type="email" 
                    name="email" 
                    id=""  
                    placeholder="Tu email" 
                    className="auth__input"
                    autoComplete="off" 
                    value={email}
                    onChange={handleInputChange}
                />

                <input 
                    type="password" 
                    name="password" 
                    id=""  
                    placeholder="Tu password" 
                    className="auth__input"
                    value={password}
                    onChange={handleInputChange}
                />

                <button 
                    type="submit"
                    className="btn btn-primary btn-block"
                    disabled={loading}
                    
                >
                    Login
                </button>

                <div className="auth__social-networks">
                    <p>Login With Social Networks</p>

                    <div 
                        className="google-btn"
                        onClick={handleGoogleSigning}
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text" >
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link to = {"/auth/register"} className="link">Register</Link>
            </form>

        </>
    )
}
