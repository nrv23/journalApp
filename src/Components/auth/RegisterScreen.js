import React from 'react'
import { Link } from 'react-router-dom';
import { setError, unSetError } from '../../actions/ui';
import useForm from '../../hooks/useForm';
import { useSelector,useDispatch } from 'react-redux';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen = () => {

   const dispatch = useDispatch()
    const { msgError} = useSelector(state => state.ui); //seleccionar un state del state global

    const [formValues,handleInputChange ] = useForm({
        name: '',
        email: '',
        password: '',
        confirm_password: ''
    });

    const { name,email,password,confirm_password } = formValues;
    const handleSubmit = e => {

        e.preventDefault();

        if(isFormValid()) {
            dispatch(startRegisterWithEmailPasswordName(email,password,name));
        }
    }

    const isFormValid = () => {

        const regex = /^([\da-z_\.-]+)@([\da-z\.-]+)\.([a-z\.]{2,6})$/; // probar en el proyecto del brete

        if(name.trim().length === 0) {
            dispatch(setError('Nombre es requerido'));
            return false;
        } else if(email.trim().length === 0 || !regex.test(email)) {
            dispatch(setError('Email no es válido'));

            return false;
        } else if(password.trim().length === 0 || confirm_password.trim().length === 0 || password.trim() !== confirm_password.trim()){
            dispatch(setError('Contraseña no es válida')); 
            return false;
        }

        dispatch(unSetError());

        return true
        
    }

    return (
        <>
            <h3 className="auth__title">Register</h3>

            <form

                onSubmit={handleSubmit}
            >

                {

                    msgError && <div className="alert alert-danger">{msgError}</div>
                }


                <input 
                    type="email" 
                    name="email" 
                    id=""  
                    placeholder="Tu email" 
                    className="auth__input" 
                    autoComplete="off" 
                    onChange={handleInputChange}
                    value={email}
                />

                <input 
                    type="text" 
                    name="name" 
                    id=""  
                    placeholder="Tu nombre" 
                    className="auth__input" 
                    autoComplete="off" 
                    onChange={handleInputChange}
                    value={name}
                />

                <input 
                    type="password" 
                    name="password" 
                    id=""  
                    placeholder="Tu password" 
                    className="auth__input"
                    onChange={handleInputChange}
                    value={password}
                />

                <input 
                    type="password" 
                    name="confirm_password" 
                    id=""  
                    placeholder="Confirmar Password" 
                    className="auth__input"
                    onChange={handleInputChange}
                    value={confirm_password}
                />

                <button 
                    type="submit"
                    className="btn btn-primary btn-block mb-5"
                >
                    Register
                </button>
                <Link to = {"/auth/login"} className="link">Login</Link>
            </form>

        </>
    )
}
