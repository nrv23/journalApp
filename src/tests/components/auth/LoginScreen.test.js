import { mount } from 'enzyme';
import { LoginScreen } from '../../../Components/auth/LoginScreen';
import {Provider} from 'react-redux';

import configureStore from 'redux-mock-store' //ES6 modules
//const { configureStore } = require('redux-mock-store') //CommonJS
import thunk from 'redux-thunk'; // middleware para acciones asincronas
import { MemoryRouter } from 'react-router-dom';
import { startGoogleLogin,startLoginEmailPassword } from '../../../actions/auth';

jest.mock('../../../actions/auth',() => ({
    startGoogleLogin: jest.fn(),
    startLoginEmailPassword: jest.fn()
}))


const middlewares = [thunk]
const mockStore = configureStore(middlewares);
const initState = {
    auth:{ },
    ui:{
        msgError: null,
        loading: false
    }
 }
let store = mockStore(initState);
store.dispatch= jest.fn();


const wrapper = mount( // cuando da el error del router, usa el memoryRouter para simular una ruta
    <Provider store= {store}>
        <MemoryRouter> 
            <LoginScreen />
        </MemoryRouter>
    </Provider>
);

describe('Pruebas en <LoginScreen />', () => {
    
    beforeEach(() => {
        store =  mockStore(initState) ;
        jest.clearAllMocks();
    })


    test('Cargar bien el compoente', () => {

         expect(wrapper).toMatchSnapshot();
    })
    
    test('Deebe disparar la accion de startLoginScreen', () => {
        
       
        const btn = wrapper.find('.google-btn');

        btn.simulate('click');

        expect(startGoogleLogin).toHaveBeenCalled();

    })
    
    test('Debe disparar el startLogin con los respectivos argumentos', () => {
           
        const btn = wrapper.find('form');

        btn.simulate('submit',{preventDefault:() => {}}); 
        expect(startLoginEmailPassword).toHaveBeenCalled();

    })
    
})
