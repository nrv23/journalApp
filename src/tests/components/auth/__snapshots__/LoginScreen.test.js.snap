// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen /> Cargar bien el compoente 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Login
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      id=""
      name="email"
      onChange={[Function]}
      placeholder="Tu email"
      type="email"
      value=""
    />
    <input
      className="auth__input"
      id=""
      name="password"
      onChange={[Function]}
      placeholder="Tu password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block"
      disabled={false}
      type="submit"
    >
      Login
    </button>
    <div
      className="auth__social-networks"
    >
      <p>
        Login With Social Networks
      </p>
      <div
        className="google-btn"
        onClick={[Function]}
      >
        <div
          className="google-icon-wrapper"
        >
          <img
            alt="google button"
            className="google-icon"
            src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
          />
        </div>
        <p
          className="btn-text"
        >
          <b>
            Sign in with google
          </b>
        </p>
      </div>
    </div>
    <a
      className="link"
      href="/auth/register"
      onClick={[Function]}
    >
      Register
    </a>
  </form>,
]
`;
